artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/Ubuntu-VM_OperatingSystemInterface-w1/source/src/main/java/org/opentosca/artifacttemplates/ubuntuvm/UbuntuVMOperatingSystemInterfaceEndpoint.java

```diff
@@ -0,0 +1,180 @@
+        } catch (Exception e) {
+            LOG.error("Could not transfer file...", e);
+            response.setError("Could not transfer file: " + e.getMessage());
+        } finally {
+            VM.disconnect();
+        }
+
+        // Send response
+        SoapUtil.sendSoapResponse(response, InvokeResponse.class, openToscaHeaders.replyTo());
+    }
+
+    @PayloadRoot(namespace = UbuntuVMConstants.NAMESPACE_URI, localPart = "runScriptRequest")
+    public void runScript(@RequestPayload RunScriptRequest request, MessageContext messageContext) {
+        LOG.info("RunScript request received!");
+
+        OpenToscaHeaders openToscaHeaders = SoapUtil.parseHeaders(messageContext);
+        InvokeResponse response = new InvokeResponse();
+        response.setMessageID(openToscaHeaders.messageId());
+
+        VirtualMachine VM = new VirtualMachine(request.getVMIP(), request.getVMUserName(), request.getVMPrivateKey());
+        try {
+            VM.connect();
+            String command = VM.replaceHome(request.getScript());
+            String result = VM.execCommand(command);
+            response.setScriptResult(SoapUtil.encode(result));
+            LOG.info("RunScript request successful");
+        } catch (Exception e) {
+            LOG.error("Could not execute script", e);
+            response.setError("Could not execute script: " + e.getMessage());
+        } finally {
+            VM.disconnect();
+        }
+
+        SoapUtil.sendSoapResponse(response, InvokeResponse.class, openToscaHeaders.replyTo());
+    }
+
+    @PayloadRoot(namespace = UbuntuVMConstants.NAMESPACE_URI, localPart = "waitForAvailabilityRequest")
+    public void waitForAvailabilityRequest(@RequestPayload WaitForAvailabilityRequest request, MessageContext messageContext) {
+        LOG.info("WaitForAvailability request received!");
+
+        OpenToscaHeaders openToscaHeaders = SoapUtil.parseHeaders(messageContext);
+        InvokeResponse response = new InvokeResponse();
+        response.setMessageID(openToscaHeaders.messageId());
+
+        VirtualMachine VM = new VirtualMachine(request.getVMIP(), request.getVMUserName(), request.getVMPrivateKey());
+        try {
+            VM.connect();
+            LOG.info("WaitForAvailability request successful");
+            response.setWaitResult("Success");
+        } catch (Exception e) {
+            LOG.error("Could not wait for availability", e);
+            response.setError("Could not wait for availability: " + e.getMessage());
+            response.setWaitResult("Error");
+        } finally {
+            VM.disconnect();
+        }
+
+        SoapUtil.sendSoapResponse(response, InvokeResponse.class, openToscaHeaders.replyTo());
+    }
+
+    @PayloadRoot(namespace = UbuntuVMConstants.NAMESPACE_URI, localPart = "testPortBindingRequest")
+    public void testPortBinding(@RequestPayload TestPortBindingRequest request, MessageContext messageContext) {
+        // TODO: testPortBinding
```

Comment:
What about this?

Reply:
idk this management operation is defined on the type but there had been no implementation if i am not wrong